diffsims/sims/diffraction_simulation.py

```diff
@@ -44,35 +45,103 @@ class DiffractionSimulation:
         the coordinates, indices, intensities, calibration and offset.
         """
-        self._coordinates = None
-        self.coordinates = coordinates
-        self.indices = indices
-        self._intensities = None
-        self.intensities = intensities
-        self._calibration = (1.0, 1.0)
+        if coordinates.ndim == 1:
+            coordinates = coordinates[None, :]
+        if indices is None:
+            indices = np.full((coordinates.shape[0], 3), np.nan)
+        if intensities is None:
+            intensities = np.full((coordinates.shape[0]), np.nan)
+        # check here whether shapes are all the same
+        if (
+            coordinates.shape[0] == indices.shape[0] == intensities.shape[0]
+            and coordinates.ndim == indices.ndim == 2
+            and intensities.ndim == 1
+        ):
+            self._coordinates = coordinates
+            self._indices = indices
+            self._intensities = intensities
+        else:
+            raise ValueError(
+                "Coordinate, intensity, and indices lists must be of the correct and matching shape."
+            )
         self.calibration = calibration
         self.offset = offset
         self.with_direct_beam = with_direct_beam
 
+    def __len__(self):
+        return self.coordinates.shape[0]
+
+    def __getitem__(self, sliced):
+        coords = self.coordinates[sliced]
+        inds = self.indices[sliced]
+        ints = self.intensities[sliced]
+        if coords.ndim == 1:
+            coords = coords[None, :]
+            inds = inds[None, :]
+            ints = ints[None]
+        # some valid numpy slices will create invalid shapes for diffraction simulation
+        if coords.ndim > 2 or coords.shape[1] > 3 or coords.shape[1] < 2:
+            raise ValueError(f"Invalid slice: {sliced}")
+        return DiffractionSimulation(
+            coords,
+            indices=inds,
+            intensities=ints,
+            calibration=self.calibration,
+            offset=self.offset,
+            with_direct_beam=self.with_direct_beam,
+        )
+
+    def _combine_data(self, other):
+        coords = np.concatenate([self._coordinates, other._coordinates], axis=0)
+        inds = np.concatenate([self._indices, other._indices], axis=0)
+        ints = np.concatenate([self._intensities, other._intensities], axis=0)
+        return coords, inds, ints
+
+    def __add__(self, other):
+        coords, inds, ints = self._combine_data(other)
+        return DiffractionSimulation(
+            coords,
+            indices=inds,
+            intensities=ints,
+            calibration=self.calibration,
+            offset=self.offset,
+            with_direct_beam=self.with_direct_beam,
+        )
+
+    def extend(self, other):
+        """Add the diffraction spots from another DiffractionSimulation"""
+        coords, inds, ints = self._combine_data(other)
+        self._coordinates = coords
+        self._indices = inds
+        self._ints = ints
```

Comment:
> Could you achieve the same thing by providing a deepcopy() method (one-line method using standard library copy.deepcopy(self)), and just keep the extend() method?

Good idea

> Would extend() be more robust if you use the setters directly, and add some checks to them, or to _combine_data()? It looks like these two methods can easily make NumPy raise errors. Perhaps adopt the checks from __init__()?

As mentioned above, the main difference between the `_` attributes and the `@property` values is that with the latter you are not sure about the sizes.
The reason I use the `_` attributes is because otherwise an extend or addition operation will remove the direct beam if it is there in the original array and `with_direct_beam` is `False`. In my opinion, if one adds information, one should not throw away data, even if it is hidden. If the user changes the `with_direct_beam` property afterwards, the direct beam will be there.

Reply:
I agree, data should be kept.